src/sympc/protocol/aby3/aby3.py

```diff
@@ -277,22 +294,136 @@ def bit_injection(x: MPCTensor, session: Session, ring_size: int) -> MPCTensor:
+            map(lambda x: x[0].resolve_pointer_type(), decomposed_shares)
+        ):
+            x1_share.append(share[0].resolve_pointer_type())
+            x2_share.append(share[1].resolve_pointer_type())
+            x3_share.append(share[2].resolve_pointer_type())
 
-        x1 = MPCTensor(shares=x1_sh, session=session, shape=x.shape)
-        x2 = MPCTensor(shares=x2_sh, session=session, shape=x.shape)
-        x3 = MPCTensor(shares=x3_sh, session=session, shape=x.shape)
+        x1 = MPCTensor(shares=x1_share, session=session, shape=x.shape)
+        x2 = MPCTensor(shares=x2_share, session=session, shape=x.shape)
+        x3 = MPCTensor(shares=x3_share, session=session, shape=x.shape)
 
         arith_share = x1 ^ x2 ^ x3
 
         return arith_share
+
+    @staticmethod
+    def full_adder(
+        a: List[MPCTensor], b: List[MPCTensor], session: Session
+    ) -> List[MPCTensor]:
+        """Perform bit addition on MPCTensors using a full adder.
+
+        Args:
+            a (List[MPCTensor]): MPCTensor with shares of bit.
+            b (List[MPCTensor]): MPCTensor with shares of bit.
+            session (Session): session the share belongs to.
+
+        Returns:
+            result (List[MPCTensor]): Result of the operation.
+
+        TODO: Should modify ripple carry adder to parallel prefix adder,currently unused.
+        """
+        ring_size = session.ring_size
+        ring_bits = get_nr_bits(ring_size)
+        c = 0  # carry bits of addition.
+        result: List[MPCTensor] = []
+        for idx in range(ring_bits):
+            s = a[idx] + b[idx] + c
+            c = a[idx] * b[idx] + c * (a[idx] + b[idx])
+            result.append(s)
+        return result
+
+    @staticmethod
+    def bit_decomposition(x: MPCTensor, session: Session) -> List[MPCTensor]:
+        """Perform ABY3 bit decomposition for conversion of arithmetic share to binary share.
+
+        Args:
+            x (MPCTensor): Arithmetic shares of secret.
+            session (Session): session the share belongs to.
+
+        Returns:
+            bin_share (List[MPCTensor]): Returns binary shares of each bit of the secret.
+
+        TODO : Should be modified to use parallel prefix adder when multiprocessing
+        functionality is integrated,currently unused.
+        """
+        x1: List[MPCTensor] = []  # bit shares of shares
+        x2: List[MPCTensor] = []
+        x3: List[MPCTensor] = []
+
+        args = [[share, "2", True] for share in x.share_ptrs]
+
+        decomposed_shares = parallel_execution(
+            ABY3.local_decomposition, session.parties
+        )(args)
+
+        # Initially we have have List[p1,p2,p3] where p1,p2,p3 are list returned from parties.
+        # Each of p1,p2,p3 is List[ [x1,x2,x3] ,...] in bit length of the session ring size.
+        # Each element of the list is a share of the shares for each bit.
+        x_sh = itertools.starmap(zip, zip(*decomposed_shares))
+
+        for share in x_sh:
```

Comment:
NITPICK: Why not directly:
`for x1_sh, x2_sh, x3_sh in x_sh`

Reply:
Ex : When we operate in session ring size, we have 64 iterators, using directly won't work george

Reply:
Modified